src/database/templates.rs

```diff
@@ -34,19 +36,70 @@ impl From<&Template> for TemplateDetails {
 pub enum TemplateError {
     AuthError(String),
     ListError(Error<ProjectsTemplatesListError>),
+    RetrieveApi(Error<ProjectsTemplatesRetrieveError>),
+    EvaluateApi(Error<ProjectsTemplatePreviewCreateError>),
+    EvaluateFailed(Vec<String>),
 }
 
 impl fmt::Display for TemplateError {
     fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
         match self {
             TemplateError::AuthError(msg) => write!(f, "Not Authenticated: {}", msg),
             TemplateError::ListError(e) => write!(f, "{}", e.to_string()),
+            TemplateError::RetrieveApi(e) => write!(f, "{}", e.to_string()),
+            TemplateError::EvaluateApi(e) => write!(f, "{}", e.to_string()),
+            TemplateError::EvaluateFailed(reasons) => {
+                write!(f, "Evaluation failed:\n  {}", reasons.join("\n  "))
+            }
         }
     }
 }
 
 impl error::Error for TemplateError {}
 
+fn unquote(orig: &str) -> String {
+    orig.trim_start_matches('"')
+        .trim_end_matches('"')
+        .to_string()
+}
+
+fn extract_reasons(content: &str) -> Vec<String> {
+    let mut reasons: Vec<String> = vec![];
+    let json_result: Result<serde_json::Value, serde_json::Error> = serde_json::from_str(content);
+    if let Ok(data) = json_result {
+        if let Some(details) = data.get("detail") {
+            if details.is_string() {
+                reasons.push(unquote(&details.to_string()))
+            } else if details.is_array() {
+                for failure in details.as_array().unwrap() {
```

Comment:
If the template APIs return a 422, there is a structured api response type which can be used as a class for serdes.

Reply:
Good call -- I will need to address that after I get a different set of API changes merged.